### PR TITLE
CLI supports to bootstrap tool package

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
CLI supports to bootstrap tool package

pf tool init --help:
image

Create package tool:
pf tool init --package xxx --tool xxx
image

package structure:
image

list tools meta in the package:
```
{
   "package_name.tool_name.tool_name":{
      "name":"tool_name",
      "type":"python",
      "inputs":{
         "connection":{
            "type":[
               "CustomConnection"
            ]
         },
         "input_text":{
            "type":[
               "string"
            ]
         }
      },
      "description":"This is tool_name tool",
      "module":"package_name.tool_name",
      "function":"tool_name"
   }
}
```

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_cli/data/package_tool/setup.py.jinja2 [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from setuptools import find_packages, setup
+
+PACKAGE_NAME = "{{ package_name }}"
+
+setup(
+    name=PACKAGE_NAME,
+    version="0.0.1",
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_cli/data/package_tool/setup.py.jinja2:7]: so when user use this CLI, the version will always be `0.0.1`? will we support custom this?

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 2]
```diff
@@ -800,6 +801,53 @@ def get_local_connections_from_executable(executable):
...
+        return anno
+    # Optional[Type] is Union[Type, NoneType], filter NoneType out
+    args = [arg for arg in get_args(anno) if arg != type(None)]  # noqa: E721
+    return args[0] if len(args) == 1 else args
+
+
+def param_to_definition(param, value_type) -> (InputDefinition, bool):
+
+    def value_to_str(val):
+        if val is inspect.Parameter.empty:
```
Comment (crazygao) [inline src/promptflow/promptflow/_sdk/_utils.py:817]: not sure if this method covered all supported values, double confirm and create all possible tests.

File: src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 
 logger = logging.getLogger(LOGGER_NAME)
 TEMPLATE_PATH = Path(__file__).resolve().parent.parent / "data" / "entry_flow"
+TOOL_TEMPLATE_PATH = Path(__file__).resolve().parent.parent / "data" / "package_tool"
```
Comment (crazygao) [inline src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py:18]: does package tool need __init__.py

Comment (crazygao) [reply src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py:18]: \_\_init\_\_.py

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py [hunk 4]
```diff
@@ -1210,6 +1211,32 @@ def mocked_connection_get(*args, **kwargs):
         outerr = capsys.readouterr()
         assert not outerr.err
 
+    def test_tool_init(self):
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py:1214]: Can we add some invalid test case? like no tools in py, or wrong tool name/ package name etc.

File: src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py (hunk 3 above)
Comment (0mza987) [inline src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py:18]: `.resolve()` seems useless when followed with `.parent`?

File: src/promptflow/promptflow/_core/tool.py [hunk 5]
```diff
@@ -39,27 +39,46 @@ def active_instance(cls) -> Optional["ToolInvoker"]:
...
+
+        new_f.__original_function = func
+        func.__wrapped_function = new_f
+        new_f.__tool = None  # This will be set when generating the tool definition.
+        new_f.__name = name
+        new_f.__description = description
+        new_f.__type = type
+        return new_f
+
+    # enable use decorator without "()" if all arguments are default values
```
Comment (0mza987) [inline src/promptflow/promptflow/_core/tool.py:78]: May need test case to guard this behavior

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        description="Manage tools for promptflow.",
+        help="pf tool",
+    )
+    subparsers = tool_parser.add_subparsers()
+    add_parser_init_tool(subparsers)
+    tool_parser.set_defaults(action="tool")
+
+
+def add_parser_init_tool(subparsers):
+    """Add tool init parser to the pf tool subparsers."""
```
Comment (0mza987) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:37]: Will we support tool init from remote data like git repo?

File: src/promptflow/promptflow/_cli/data/package_tool/tool.py.jinja2 [hunk 7]
```diff
@@ -0,0 +1,11 @@
+from promptflow import tool
+from promptflow.connections import CustomConnection
+
+
+@tool(name="{{ tool_name }}", description="This is {{ tool_name }} tool")
+def {{ tool_name }}(connection: CustomConnection, input_text: str) -> str:
+    # Replace with your tool code.
```
Comment (0mza987) [inline src/promptflow/promptflow/_cli/data/package_tool/tool.py.jinja2:7]: So after the init, user needs to manually replace the code here? I thought we can generate it for user.

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 8]
```diff
@@ -37,6 +38,7 @@ def __init__(self):
         # Lazy init to avoid azure credential requires too early
         self._connections = None
         self._flows = FlowOperations()
+        self._tools = ToolOperations()
```
Comment (0mza987) [inline src/promptflow/promptflow/_sdk/_pf_client.py:41]: Do we have sdk interface for tool init?

File: src/promptflow/promptflow/_core/tool.py (hunk 5 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_core/tool.py:78]: There has been added to the test case that use decorator without bracket.
image

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 6 above)
Comment (0mza987) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:37]: I see, I mis-understood it with a aml's  `generate package`

File: src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py (hunk 4 above)
Comment (lalala123123) [reply src/promptflow/tests/sdk_cli_test/e2etests/test_cli.py]: Added test case to check the validation of package/tool name.

File: src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_init_entry_generators.py]: remove resolve()

**Review by 0mza987 — APPROVED** [review]

**Review by crazygao — APPROVED** [review]

**Pull request merged**